libcaf_io/src/routing_table.cpp

```diff
@@ -34,122 +34,99 @@ routing_table::~routing_table() {
+    return *i->second.begin();
+  return none;
 }
 
-void routing_table::erase_direct(const connection_handle& hdl,
-                                 erase_callback& cb) {
+node_id routing_table::erase_direct(const connection_handle& hdl) {
+  std::unique_lock<std::mutex> guard{mtx_};
   auto i = direct_by_hdl_.find(hdl);
   if (i == direct_by_hdl_.end())
-    return;
-  cb(i->second);
-  parent_->parent().notify<hook::connection_lost>(i->second);
+    return none;
   direct_by_nid_.erase(i->second);
+  node_id result = std::move(i->second);
   direct_by_hdl_.erase(i->first);
+  return result;
 }
 
 bool routing_table::erase_indirect(const node_id& dest) {
+  std::unique_lock<std::mutex> guard{mtx_};
   auto i = indirect_.find(dest);
   if (i == indirect_.end())
     return false;
-  if (parent_->parent().has_hook())
-    for (auto& nid : i->second)
-      parent_->parent().notify<hook::route_lost>(nid, dest);
   indirect_.erase(i);
   return true;
 }
 
 void routing_table::add_direct(const connection_handle& hdl,
                                const node_id& nid) {
-  CAF_ASSERT(direct_by_hdl_.count(hdl) == 0);
-  CAF_ASSERT(direct_by_nid_.count(nid) == 0);
-  direct_by_hdl_.emplace(hdl, nid);
-  direct_by_nid_.emplace(nid, hdl);
-  parent_->parent().notify<hook::new_connection_established>(nid);
+  std::unique_lock<std::mutex> guard{mtx_};
+  auto hdl_added = direct_by_hdl_.emplace(hdl, nid).second;
+  auto nid_added = direct_by_nid_.emplace(nid, hdl).second;
+  CAF_ASSERT(hdl_added && nid_added);
+  CAF_IGNORE_UNUSED(hdl_added);
+  CAF_IGNORE_UNUSED(nid_added);
 }
 
 bool routing_table::add_indirect(const node_id& hop, const node_id& dest) {
-  auto i = blacklist_.find(dest);
-  if (i == blacklist_.end() || i->second.count(hop) == 0) {
-    auto& hops = indirect_[dest];
-    auto added_first = hops.empty();
-    hops.emplace(hop);
-    parent_->parent().notify<hook::new_route_added>(hop, dest);
-    return added_first;
-  }
-  return false; // blacklisted
-}
-
-bool routing_table::reachable(const node_id& dest) {
-  return direct_by_nid_.count(dest) > 0 || indirect_.count(dest) > 0;
-}
-
-size_t routing_table::erase(const node_id& dest, erase_callback& cb) {
-  cb(dest);
-  size_t res = 0;
-  auto i = indirect_.find(dest);
-  if (i != indirect_.end()) {
-    res = i->second.size();
-    for (auto& nid : i->second) {
-      cb(nid);
-      parent_->parent().notify<hook::route_lost>(nid, dest);
-    }
-    indirect_.erase(i);
-  }
-  auto hdl = lookup_direct(dest);
-  if (hdl) {
-    direct_by_hdl_.erase(*hdl);
-    direct_by_nid_.erase(dest);
-    parent_->parent().notify<hook::connection_lost>(dest);
-    ++res;
-  }
-  return res;
+  std::unique_lock<std::mutex> guard{mtx_};
+  // Never add indirect entries if we already have direct connection.
+  if (direct_by_nid_.count(dest) != 0)
+    return false;
+  // Never add indirect entries if we don't have a connection to the hop.
+  if (direct_by_nid_.count(hop) == 0)
+    return false;
```

Comment:
I have conflicting views on the return value of this function. Just looking at the name, sure, it returns true if the entry was added. But this is not helpful in telling us what what the state is now. Even if it returns `false` we might be able to reach the node (line 121) or not (line 124).

Reply:
I agree, the semantics of the routing table is convoluted. However, since we already voted to rip out the entire routing stuff in `caf_net`, I'd say we don't invest further cycles here. Ok?